File: ArrayAndHashmap/leet128.js
// Longest Consecutive Sequence: Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence. You must write an algorithm that runs in O(n) time.

let nums = [100, 4, 200, 1, 3, 2]; //Output: 4 for [1,2,3,4]

var longestConsecutive = function (nums) {
  let numsSet = new Set(nums);
  let maxScore = 0;

  for (const num of [...numsSet]) {
    const prevNum = num - 1;
    if (numsSet.has(prevNum)) continue;

    let [currNum, score] = [num, 1];
    const isStreak = () => numsSet.has(currNum + 1);

    while (isStreak()) {
      currNum++;
      score++;
    }
    maxScore = Math.max(maxScore, score);
  }
  return maxScore;
};

console.log(longestConsecutive(nums));

File: ArrayAndHashmap/leet128.py
# Leet 128. Longest Consecutive Sequence
# Given an unsorted array of integers nums, return the length of the longest consecutive elements sequence.
# You must write an algorithm that runs in O(n) time.

class Solution:
  def longestConsecutiveSequence(self, nums: List[int]) -> int:
    numsSet = set(nums)
    longest = 0
    
    for n in nums:
      #checking if its the beginning of a sequence
      if (n - 1) in numsSet:
        length = 0
        while (n + length):
          length += 1
          longest = max(length, longest)
      return longest